algebird-core/src/main/scala/com/twitter/algebird/maps/increment.scala

```diff
@@ -0,0 +1,149 @@
+        else if (keyOrdering.gt(di.key, data.key)) rNode(data, lsub, rsub.inc(di))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          rNode(d, lsub, rsub)
+        }
+      } else {
+        if (keyOrdering.lt(di.key, data.key)) balance(bNode(data, lsub.inc(di), rsub))
+        else if (keyOrdering.gt(di.key, data.key)) balance(bNode(data, lsub, rsub.inc(di)))
+        else {
+          val d = new DataMap[K, V] {
+            val key = data.key
+            val value = valueMonoid.plus(data.value, di.value)
+          }
+          bNode(d, lsub, rsub)
+        }
+      }
+  }
+}
+
+import tree._
+
+object infra {
+  class Inject[K, V](val keyOrdering: Ordering[K], val valueMonoid: Monoid[V])
+    extends Serializable {
+    def iNode(clr: Color, dat: Data[K], ls: Node[K], rs: Node[K]) =
+      new Inject[K, V](keyOrdering, valueMonoid) with INodeInc[K, V] with IncrementMap[K, V] {
+        // INode[K]
+        val color = clr
+        val lsub = ls.asInstanceOf[NodeInc[K, V]]
+        val rsub = rs.asInstanceOf[NodeInc[K, V]]
+        val data = dat.asInstanceOf[DataMap[K, V]]
+      }
+  }
+}
+
+import infra._
+
+/**
+ * An inheritable and mixable trait for adding increment operation to ordered maps
+ * @tparam K The key type
+ * @tparam V The value type
+ * @tparam IN The node type of the concrete internal R/B tree subclass
+ * @tparam M The map self-type of the concrete map subclass
+ */
+trait IncrementMapLike[K, V, IN <: INodeInc[K, V], M <: IncrementMapLike[K, V, IN, M]]
+  extends NodeInc[K, V] with OrderedMapLike[K, V, IN, M] {
+
+  /**
+   * Add (w.r.t. valueMonoid) a given value to the value currently stored at key.
+   * @note If key is not present, equivalent to insert(k, valueMonoid.plus(valueMonoid.zero, iv)
+   */
+  def increment(k: K, iv: V) = this.incr(
+    new DataMap[K, V] {
+      val key = k
+      val value = iv
+    }).asInstanceOf[M]
+}
+
+sealed trait IncrementMap[K, V] extends IncrementMapLike[K, V, INodeInc[K, V], IncrementMap[K, V]] {
+  override def toString =
+    "IncrementMap(" +
+      nodesIterator.map(n => s"${n.data.key} -> ${n.data.value}").mkString(", ") +
+      ")"
+}
+
+object IncrementMap {
+  /**
+   * Instantiate a new empty IncrementMap from key and value types
+   * {{{
+   * import scala.language.reflectiveCalls
+   * import com.twitter.algebird.maps.increment._
+   *
+   * // map strings to integers, using default string ordering and default value monoid
+   * val map1 = IncrementMap.key[String].value[Int]
+   * // Use a custom ordering
+   * val ord: Ordering[String] = ...
+   * val map2 = IncrementMap.key(ord).value[Int]
+   * // A custom value monoid, defines what 'increment' means
+   * val mon: Monoid[Int] = ...
+   * val map2 = IncrementMap.key[String].value(mon)
+   * }}}
+   */
+  def key[K](implicit ord: Ordering[K]) = new AnyRef {
```

Comment:
you are using several types here, why not use a named type here? `UnsetValueIncrementMap` or something.